File: app/src/main/java/vergecurrency/vergewallet/helpers/utils/LanguagesUtils.java
package vergecurrency.vergewallet.helpers.utils;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;

import com.akexorcist.localizationactivity.core.LocalizationActivityDelegate;
import com.google.gson.GsonBuilder;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Locale;

import vergecurrency.vergewallet.Constants;
import vergecurrency.vergewallet.service.model.Language;
import vergecurrency.vergewallet.service.model.PreferencesManager;

public final class LanguagesUtils {

	public LanguagesUtils() {

	}


	public static ArrayList<Language> loadLanguagesFromFile(Context c) {
		JSONParser parser = new JSONParser();
		ArrayList<Language> languages;
		try {
			InputStream is = c.getAssets().open(Constants.LANGUAGES_FILE_PATH);
			InputStreamReader isr = new InputStreamReader(is);
			JSONObject jsonObject = (JSONObject) parser.parse(isr);
			JSONArray languagesJSON = (JSONArray) jsonObject.get("languages");


			Language[] languagesArray;
			languagesArray = new GsonBuilder().create().fromJson(languagesJSON.toJSONString(), Language[].class);
			languages = new ArrayList<>(Arrays.asList(languagesArray));


		} catch (IOException e) {
			languages = null;
			e.printStackTrace();
		} catch (ParseException e) {
			e.printStackTrace();
			languages = null;
		}

		return languages;
	}

	public static Context onAttach(Context context) {
		String lang = getPersistedData();
		return setLocale(context, lang);
	}

	public static String getLanguage() {
		return getPersistedData();
	}

	public static Context setLocale(Context context, String lang) {
		String language = Language.getLanguageFromJson(lang).getLanguage();
		persist(lang);

		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
			return updateResources(context, language);
		}

		return updateResourcesLegacy(context, language);
	}

	private static String getPersistedData() {
		PreferencesManager pm = PreferencesManager.getInstance();
		return pm.getCurrentLanguage();
	}

	private static void persist(String language) {
		PreferencesManager pm = PreferencesManager.getInstance();
		pm.setCurrentLanguage(language);
	}

	@TargetApi(Build.VERSION_CODES.N)
	private static Context updateResources(Context context, String language) {
		Locale locale = new Locale(language);
		Locale.setDefault(locale);

		Configuration configuration = context.getResources().getConfiguration();
		configuration.setLocale(locale);

		return context.createConfigurationContext(configuration);
	}

	@SuppressWarnings("deprecation")
	private static Context updateResourcesLegacy(Context context, String language) {
		Locale locale = new Locale(language);
		Locale.setDefault(locale);

		Resources resources = context.getResources();

		Configuration configuration = resources.getConfiguration();
		configuration.locale = locale;

		resources.updateConfiguration(configuration, resources.getDisplayMetrics());

		return context;
	}


}

File: app/src/main/kotlin/vergecurrency/vergewallet/helpers/utils/LanguagesUtils.kt
package vergecurrency.vergewallet.helpers.utils

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import com.google.gson.GsonBuilder
import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import org.json.simple.parser.ParseException
import vergecurrency.vergewallet.Constants
import vergecurrency.vergewallet.service.model.Language
import vergecurrency.vergewallet.service.model.PreferencesManager
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

object LanguagesUtils {

    fun loadLanguagesFromFile(c: Context): ArrayList<Language>? {
        val parser = JSONParser()
        var languages: ArrayList<Language>?
        try {
            val `is` = c.assets.open(Constants.LANGUAGES_FILE_PATH)
            val isr = InputStreamReader(`is`)
            val jsonObject = parser.parse(isr) as JSONObject
            val languagesJSON = jsonObject["languages"] as JSONArray?


            val languagesArray: Array<Language>
            languagesArray = GsonBuilder().create().fromJson(languagesJSON!!.toJSONString(), Array<Language>::class.java)
            languages = ArrayList(Arrays.asList(*languagesArray))


        } catch (e: IOException) {
            languages = null
            e.printStackTrace()
        } catch (e: ParseException) {
            e.printStackTrace()
            languages = null
        }

        return languages
    }

    fun onAttach(context: Context): Context {
        val lang = persistedData
        return setLocale(context, lang)
    }

    val language: String
        get() = persistedData

    fun setLocale(context: Context, lang: String): Context {
        val language = Language.getLanguageFromJson(lang).language
        persist(lang)

        return updateResources(context, language!!)
    }

    private val persistedData: String
        get() = PreferencesManager.currentLanguage!!

    private fun persist(language: String) {
        PreferencesManager.currentLanguage = language
    }

    @TargetApi(Build.VERSION_CODES.N)
    private fun updateResources(context: Context, language: String): Context {
        val locale = Locale(language)
        Locale.setDefault(locale)

        val configuration = context.resources.configuration
        configuration.setLocale(locale)

        return context.createConfigurationContext(configuration)
    }
}
